components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts

```diff
@@ -0,0 +1,188 @@
+import {
+  Component,
+  OnInit,
+  ViewChild,
+  Input,
+  Output,
+  EventEmitter
+} from '@angular/core';
+
+import { FormBuilder, Validators, FormGroup } from '@angular/forms';
+import { Regex } from 'app/helpers/auth/regex';
+
+import { MatSelect } from '@angular/material/select';
+import { DynamicGrid } from '../create-environment-modal/create-environment-modal.component';
+
+@Component({
+  selector: 'app-infra-environment-constraint',
+  templateUrl: './infra-environment-constraint.component.html',
+  styleUrls: ['./infra-environment-constraint.component.scss']
+})
+
+export class InfraEnvironmentConstraintComponent implements OnInit {
+
+  @Input() constraintKeys: string[] = [];
+  @Input() dynamicArray: Array<DynamicGrid> = [];
+  @Input() name_id: string;
+  @Output() constraintValues: EventEmitter<DynamicGrid[]> = new EventEmitter();
+  @ViewChild(MatSelect) select: MatSelect;
+
+  public conflictError = false;
+  public operator_id = '';
+  public operatorKeys: string[] = [];
+  public secondFormGroup: FormGroup;
+  public selectedLangs: string[] = [];
+  public rowKeys: string[] = [];
+
+  constructor(
+    private fb: FormBuilder
+  ) {
+    this.secondFormGroup = this.fb.group({
+      version: ['', [Validators.required, Validators.pattern(Regex.patterns.VALID_VERSION)]]
+    });
+
+    this.operatorKeys = ['~>', '>=', '>', '=', '<', '<='];
+
+    this.operator_id = this.operatorKeys[0];
+  }
+
+  ngOnInit() {
+
+    this.conflictError = false;
+    this.constraintKeys.forEach(element =>
+
+      this.rowKeys.push(element)
+    );
+  }
+
+  selected(value, i) {
```

Comment:
Wow! What is this method doing?
Suggest each block should be extracted into a named function so this function will look something like:
```
selectLanguage(...) {
  setLanguageValueFromIndex(...);
  setNameIn???ArraytoValue(...);
  doSomethingSomethingWithConstraints(...);
  etc...
}
```
And looking further, you can then reuse those sub-functions below in the `handle*` methods.

Reply:
working on these changes, will update

Reply:
Those **handle** methods are specific to name, operator, version that's why i created separated functions to handle these changes.

Reply:
Looks like all you did was rename `selected` to `handleName`, etc., which is not at all what I was asking. Inside `handleEditName` you have what looks like 4 blocks of code. Each of those 4 blocks should be extracted into a named function so `handleEditName` will look something like I showed above. _And then you can reuse those subfunctions in `handleEditOperator` and `handleEditVersion`._

Reply:
I have added the functions for each four blocks.